File: mod/orbit/openarena/openarena-rs/src/mcp.rs
//! MCP server core — JSON-RPC 2.0 dispatch shared by the Streamable HTTP
//! endpoint (/mcp) and stdio mode (--stdio).
//!
//! Every REST route funnels through `call_tool` too, so an arena capability is
//! defined exactly once: what an agent can do over MCP is what a browser can
//! do over HTTP, always.

use crate::arena;
use crate::bench;
use crate::judge;
use crate::players;
use serde_json::{json, Value};
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};

pub const PROTOCOL_VERSION: &str = "2025-06-18";
pub const SERVER_NAME: &str = "openarena";
pub const SERVER_VERSION: &str = env!("CARGO_PKG_VERSION");

pub fn tool_list() -> Value {
    json!([
        {
            "name": "arena_info",
            "description": "What this arena is and what is in it: task, competitor and match counts, the languages the judge runs, and the competitor kinds it can drive.",
            "inputSchema": { "type": "object", "properties": {} }
        },
        {
            "name": "list_tasks",
            "description": "Every task in the arena, newest last. Filter by free text, tag or kind.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "q": { "type": "string", "description": "Match title, slug or statement" },
                    "tag": { "type": "string" },
                    "kind": { "type": "string", "description": "e.g. coding" }
                }
            }
        },
        {
            "name": "get_task",
            "description": "One task in full: the statement, the language and mode, and the visible test cases. Hidden cases show only their name and weight.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "task": { "type": "string", "description": "Task id or slug" },
                    "reveal": { "type": "boolean", "default": false, "description": "Include hidden cases — for the task author, not for entrants" }
                },
                "required": ["task"]
            }
        },
        {
            "name": "create_task",
            "description": "Upload a task. `io` mode feeds each case's stdin to the program and compares stdout with `expect`; `unit` mode runs each case's `program`, which imports the submission and passes by exiting 0. Mark the cases you do not want entrants to see as hidden.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "title": { "type": "string" },
                    "statement": { "type": "string", "description": "What the program must do, in prose" },
                    "slug": { "type": "string", "description": "Defaults to a slug of the title" },
                    "kind": { "type": "string", "default": "coding" },
                    "language": { "type": "string", "default": "any", "description": "any | python | javascript | bash" },
                    "mode": { "type": "string", "enum": ["io", "unit"], "default": "io" },
                    "starter": { "type": "string", "description": "Optional starter code shown to entrants" },
                    "timeout_ms": { "type": "integer", "default": 10000, "description": "Wall clock per case" },
                    "tags": { "type": "array", "items": { "type": "string" } },
                    "author": { "type": "string" },
                    "tests": {
                        "type": "array",
                        "description": "The grading contract — at least one case",
                        "items": {
                            "type": "object",
                            "properties": {
                                "name": { "type": "string" },
                                "stdin": { "type": "string", "description": "io mode: fed to the program" },
                                "expect": { "type": "string", "description": "io mode: the required stdout" },
                                "compare": { "type": "string", "enum": ["trim", "exact", "contains"], "default": "trim" },
                                "program": { "type": "string", "description": "unit mode: the grader source, run in the same directory as the submission" },
                                "args": { "type": "array", "items": { "type": "string" } },
                                "hidden": { "type": "boolean", "default": false },
                                "weight": { "type": "number", "default": 1 }
                            }
                        }
                    }
                },
                "required": ["title", "tests"]
            }
        },
        {
            "name": "delete_task",
            "description": "Remove a task from the arena. Past matches keep their record.",
            "inputSchema": {
                "type": "object",
                "properties": { "task": { "type": "string" } },
                "required": ["task"]
            }
        },
        {
            "name": "list_agents",
            "description": "Every competitor entered in the arena, strongest first.",
            "inputSchema": { "type": "object", "properties": {} }
        },
        {
            "name": "enter_agent",
            "description": "Enter a competitor. `agent_mod` runs an agent from this fleet's agent module (config: base, agent, model, prompt, toolbox, steps, free, key). `http` posts the task to config.url and reads a program back. `ap` drives an Agent Protocol v1 server at config.base. `static` submits config.code every time — a baseline to measure against.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "name": { "type": "string" },
                    "kind": { "type": "string", "enum": ["agent_mod", "http", "ap", "static"], "default": "agent_mod" },
                    "config": { "type": "object", "description": "Driver settings — see the description" },
                    "owner": { "type": "string", "description": "Who entered it" },
                    "note": { "type": "string" }
                },
                "required": ["name"]
            }
        },
        {
            "name": "remove_agent",
            "description": "Withdraw a competitor. Past matches keep their record.",
            "inputSchema": {
                "type": "object",
                "properties": { "agent": { "type": "string", "description": "Competitor id or name" } },
                "required": ["agent"]
            }
        },
        {
            "name": "run_match",
            "description": "Put competitors on a task. They all get the same brief at the same moment, answer concurrently, and every answer is graded against the same cases. Two or more entrants makes the match rated and moves Elo; one entrant is practice.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "task": { "type": "string", "description": "Task id or slug" },
                    "agents": { "type": "array", "items": { "type": "string" }, "description": "Competitor ids or names" }
                },
                "required": ["task", "agents"]
            }
        },
        {
            "name": "submit",
            "description": "Grade a program against a task without a match — how an outside agent or a human plays. Records the submission and the pass rate; never moves Elo, since nobody was raced.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "task": { "type": "string" },
                    "code": { "type": "string", "description": "The complete program" },
                    "language": { "type": "string", "description": "python | javascript | bash; defaults to the task's" },
                    "agent": { "type": "string", "description": "Credit the submission to this competitor" }
                },
                "required": ["task", "code"]
            }
        },
        {
            "name": "list_matches",
            "description": "Recent matches, newest first, with the scoreboard of each.",
            "inputSchema": {
                "type": "object",
                "properties": { "limit": { "type": "integer", "default": 20 } }
            }
        },
        {
            "name": "get_match",
            "description": "One match in full: every case result and the program each competitor submitted.",
            "inputSchema": {
                "type": "object",
                "properties": { "id": { "type": "string" } },
                "required": ["id"]
            }
        },
        {
            "name": "leaderboard",
            "description": "Competitors ranked by Elo, with pass rate, mean score and mean time to answer.",
            "inputSchema": {
                "type": "object",
                "properties": { "limit": { "type": "integer", "default": 20 } }
            }
        },
        {
            "name": "bench_sources",
            "description": "The standardized benchmarks this arena can pull off the web — HumanEval, HumanEval+, MBPP, CodeContests — plus the three generic sources (`hf` for any HuggingFace dataset, `json` for any JSON/JSONL url, `html` to scrape a problem page). Says whether fetching is switched on.",
            "inputSchema": { "type": "object", "properties": {} }
        },
        {
            "name": "hf_search",
            "description": "Search HuggingFace for a dataset to race agents on. Answers dataset ids, downloads, likes and whether the set is gated — feed one straight to hf_dataset. This is the front door of the HuggingFace interface: you do not have to already know the name of a benchmark to import it.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "query": { "type": "string", "description": "Free text, matched against dataset ids and cards — e.g. `code benchmark`, `humaneval`, `sql`" },
                    "limit": { "type": "integer", "default": 20, "description": "1–100" },
                    "sort": { "type": "string", "enum": ["downloads", "likes", "recent"], "default": "downloads" },
                    "filter": { "type": "string", "description": "A hub filter tag, e.g. `task_categories:text-generation` or `language:en`" },
                    "author": { "type": "string", "description": "Only this owner's datasets, e.g. `openai`" },
                    "refresh": { "type": "boolean", "default": false }
                }
            }
        },
        {
            "name": "hf_dataset",
            "description": "Everything needed to import one HuggingFace dataset: its configs and splits, the columns of a real row, and the style + field map those columns imply. The `suggested` object it answers with IS the argument object for bench_preview — call this, then pass `suggested` through, and a dataset nobody wrote an adapter for becomes arena tasks.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "dataset": { "type": "string", "description": "e.g. openai/openai_humaneval" },
                    "config": { "type": "string", "description": "Defaults to the first config the viewer lists" },
                    "split": { "type": "string", "description": "Defaults to the held-out split — test, then validation, then train" },
                    "refresh": { "type": "boolean", "default": false }
                },
                "required": ["dataset"]
            }
        },
        {
            "name": "bench_preview",
            "description": "Fetch a benchmark and show the arena tasks it would become — statements, cases and which of them would be hidden. Imports nothing. Always the first call: a benchmark you have not looked at converts into tasks you have not read.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "source": { "type": "string", "default": "humaneval", "description": "A source id from bench_sources" },
                    "limit": { "type": "integer", "default": 10, "description": "Records to take, 1–200" },
                    "offset": { "type": "integer", "default": 0, "description": "Where to start — how you take a benchmark a page at a time" },
                    "url": { "type": "string", "description": "json/html sources: what to fetch" },
                    "dataset": { "type": "string", "description": "hf source: e.g. openai/openai_humaneval" },
                    "config": { "type": "string" },
                    "split": { "type": "string", "default": "test" },
                    "style": { "type": "string", "enum": ["humaneval", "asserts", "io", "html"], "description": "How a record becomes a task — defaults to the source's" },
                    "map": { "type": "object", "description": "Field paths per part of the task: title, statement, starter, entry_point, program, asserts, imports, inputs, outputs, hidden_inputs, hidden_outputs. Dotted paths walk into nested objects. Merged over the style's defaults." },
                    "hide_after": { "type": "integer", "description": "Keep this many cases visible and hide the rest — overrides the style's own split" },
                    "max_cases": { "type": "integer", "default": 12, "description": "Cap on cases per task; visible ones are kept first" },
                    "timeout_ms": { "type": "integer", "default": 10000 },
                    "language": { "type": "string", "description": "io tasks only — unit tasks are Python by construction" },
                    "split_asserts": { "type": "boolean", "default": true, "description": "Cut an all-assertion grader into one case per assertion, so a near-miss scores partial credit" },
                    "tags": { "type": "array", "items": { "type": "string" } },
                    "slug_prefix": { "type": "string", "description": "Defaults to the source id" },
                    "refresh": { "type": "boolean", "default": false, "description": "Ignore the day-long fetch cache" }
                }
            }
        },
        {
            "name": "bench_import",
            "description": "Convert a benchmark into arena tasks and keep them. Same arguments as bench_preview. A slug already in the arena is skipped, not an error, so paging through a benchmark with `offset` is safe to repeat; the reply carries `next_offset` for the next page.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "source": { "type": "string", "default": "humaneval" },
                    "limit": { "type": "integer", "default": 10 },
                    "offset": { "type": "integer", "default": 0 },
                    "url": { "type": "string" },
                    "dataset": { "type": "string" },
                    "config": { "type": "string" },
                    "split": { "type": "string" },
                    "style": { "type": "string", "enum": ["humaneval", "asserts", "io", "html"] },
                    "map": { "type": "object" },
                    "hide_after": { "type": "integer" },
                    "max_cases": { "type": "integer", "default": 12 },
                    "timeout_ms": { "type": "integer", "default": 10000 },
                    "language": { "type": "string" },
                    "split_asserts": { "type": "boolean", "default": true },
                    "tags": { "type": "array", "items": { "type": "string" } },
                    "slug_prefix": { "type": "string" },
                    "refresh": { "type": "boolean", "default": false },
                    "dry_run": { "type": "boolean", "default": false, "description": "Convert and report, write nothing — bench_preview by another name" }
                }
            }
        }
    ])
}

fn s(args: &Value, key: &str) -> String {
    args.get(key)
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_string()
}

fn u(args: &Value, key: &str, default: u64) -> u64 {
    args.get(key).and_then(|v| v.as_u64()).unwrap_or(default)
}

fn b(args: &Value, key: &str) -> bool {
    args.get(key).and_then(|v| v.as_bool()).unwrap_or(false)
}

/// The one place an arena capability is implemented.
pub async fn call_tool(name: &str, args: &Value) -> Result<Value, String> {
    match name {
        "arena_info" => Ok(arena::info()),
        "list_tasks" => Ok(arena::list_tasks(args)),
        "get_task" => {
            let key = s(args, "task");
            if key.is_empty() {
                return Err("get_task requires `task`".into());
            }
            arena::get_task(&key, b(args, "reveal"))
        }
        "create_task" => arena::create_task(args),
        "delete_task" => {
            let key = s(args, "task");
            if key.is_empty() {
                return Err("delete_task requires `task`".into());
            }
            arena::delete_task(&key)
        }
        "list_agents" => Ok(arena::list_agents()),
        "enter_agent" => arena::enter_agent(args),
        "remove_agent" => {
            let key = s(args, "agent");
            if key.is_empty() {
                return Err("remove_agent requires `agent`".into());
            }
            arena::remove_agent(&key)
        }
        "run_match" => {
            let task = s(args, "task");
            if task.is_empty() {
                return Err("run_match requires `task`".into());
            }
            let agents: Vec<String> = match args.get("agents") {
                Some(Value::Array(a)) => a
                    .iter()
                    .filter_map(|v| v.as_str().map(|s| s.trim().to_string()))
                    .filter(|s| !s.is_empty())
                    .collect(),
                Some(Value::String(one)) => vec![one.trim().to_string()],
                _ => vec![],
            };
            if agents.is_empty() {
                return Err("run_match requires `agents` — at least one competitor".into());
            }
            arena::run_match(&task, &agents).await
        }
        "submit" => {
            let task = s(args, "task");
            let code = args.get("code").and_then(|v| v.as_str()).unwrap_or("");
            if task.is_empty() || code.trim().is_empty() {
                return Err("submit requires `task` and `code`".into());
            }
            let agent = s(args, "agent");
            arena::submit(
                &task,
                code,
                &s(args, "language"),
                if agent.is_empty() { None } else { Some(agent.as_str()) },
            )
            .await
        }
        "list_matches" => Ok(arena::list_matches(u(args, "limit", 20) as usize)),
        "get_match" => {
            let id = s(args, "id");
            if id.is_empty() {
                return Err("get_match requires `id`".into());
            }
            arena::get_match(&id)
        }
        "leaderboard" => Ok(arena::leaderboard(u(args, "limit", 20) as usize)),
        "bench_sources" => Ok(bench::sources_tool()),
        "hf_search" => bench::hf_search(args).await,
        "hf_dataset" => bench::hf_dataset(args).await,
        "bench_preview" => bench::preview(args).await,
        "bench_import" => bench::import(args).await,
        other => Err(format!(
            "unknown tool: {other} — this arena runs {:?} against {:?}",
            players::KINDS,
            judge::languages()
        )),
    }
}

fn rpc_result(id: Value, result: Value) -> Value {
    json!({ "jsonrpc": "2.0", "id": id, "result": result })
}

fn rpc_error(id: Value, code: i64, message: &str) -> Value {
    json!({ "jsonrpc": "2.0", "id": id, "error": { "code": code, "message": message } })
}

/// Handle one JSON-RPC message. Returns None for notifications (no reply).
pub async fn handle_message(msg: &Value) -> Option<Value> {
    let method = msg.get("method").and_then(|m| m.as_str()).unwrap_or("");
    let params = msg.get("params").cloned().unwrap_or_else(|| json!({}));

    let id = match msg.get("id").cloned() {
        Some(id) if !id.is_null() => id,
        _ => return None,
    };

    Some(match method {
        "initialize" => rpc_result(
            id,
            json!({
                "protocolVersion": params.get("protocolVersion").and_then(|v| v.as_str()).unwrap_or(PROTOCOL_VERSION),
                "capabilities": { "tools": {} },
                "serverInfo": { "name": SERVER_NAME, "version": SERVER_VERSION }
            }),
        ),
        "ping" => rpc_result(id, json!({})),
        "tools/list" => rpc_result(id, json!({ "tools": tool_list() })),
        "tools/call" => {
            let name = params.get("name").and_then(|n| n.as_str()).unwrap_or("");
            let args = params.get("arguments").cloned().unwrap_or_else(|| json!({}));
            match call_tool(name, &args).await {
                Ok(v) => rpc_result(
                    id,
                    json!({
                        "content": [{ "type": "text", "text": serde_json::to_string(&v).unwrap_or_default() }],
                        "structuredContent": v,
                        "isError": false
                    }),
                ),
                Err(e) => rpc_result(
                    id,
                    json!({ "content": [{ "type": "text", "text": e }], "isError": true }),
                ),
            }
        }
        "resources/list" => rpc_result(id, json!({ "resources": [] })),
        "prompts/list" => rpc_result(id, json!({ "prompts": [] })),
        _ => rpc_error(id, -32601, &format!("method not found: {method}")),
    })
}

/// stdio transport: newline-delimited JSON-RPC on stdin/stdout.
/// Usage: openarena-api --stdio
pub async fn run_stdio() {
    let stdin = BufReader::new(tokio::io::stdin());
    let mut stdout = tokio::io::stdout();
    let mut lines = stdin.lines();
    while let Ok(Some(line)) = lines.next_line().await {
        let line = line.trim().to_string();
        if line.is_empty() {
            continue;
        }
        let msg: Value = match serde_json::from_str(&line) {
            Ok(v) => v,
            Err(e) => {
                let err = rpc_error(Value::Null, -32700, &format!("parse error: {e}"));
                let _ = stdout.write_all(format!("{err}\n").as_bytes()).await;
                let _ = stdout.flush().await;
                continue;
            }
        };
        if let Some(resp) = handle_message(&msg).await {
            let _ = stdout.write_all(format!("{resp}\n").as_bytes()).await;
            let _ = stdout.flush().await;
        }
    }
}

File: mod/orbit/skills/skillsrc/mcp.py
"""mcp — the same market, as a Model Context Protocol server.

Ten tools, each one a method on Market. An MCP client and a curl get the same
answers because there is one implementation behind both; this file is the
tool schema and the dispatch, and nothing else.

Streamable HTTP lives in api.py (POST /mcp); this module is transport-free so
a stdio client can import it directly.
"""
import json
from typing import Any, Dict, List

from .market import Market
from .sources import SOURCE_IDS

PROTOCOL = "2025-06-18"
SERVER = {"name": "skills", "version": "0.1.0",
          "title": "Skills — the agent skill marketplace"}

_STR = {"type": "string"}
_INT = {"type": "integer"}
_BOOL = {"type": "boolean"}
_LIST = {"type": "array", "items": {"type": "string"}}

TOOLS: List[Dict[str, Any]] = [
    {"name": "skills_search",
     "description": ("Scan the open web for agent skills. One query fans out to the "
                     "official anthropics/skills catalog, GitHub repo search, skill "
                     "topics, GitHub code search for SKILL.md files, curated "
                     "awesome-lists and the modules on this host; duplicates merge "
                     "and results are ranked. Returns cards, not documents — open one "
                     "with skills_get."),
     "inputSchema": {"type": "object", "properties": {
         "q": dict(_STR, description="what the skill should do, e.g. 'pdf forms'"),
         "sources": dict(_LIST, description=f"limit to some of: {', '.join(SOURCE_IDS)}"),
         "limit": dict(_INT, description="max results (default 30)"),
         "fresh": dict(_BOOL, description="bypass the cache")}}},
    {"name": "skills_sources",
     "description": "The sources a scan reaches, and which are ready (code search needs a token).",
     "inputSchema": {"type": "object", "properties": {}}},
    {"name": "skills_get",
     "description": ("Open one search result and read the actual SKILL.md behind it. "
                     "Takes a result id (gh:owner/repo[:path], mod:orbit/name, or a URL) "
                     "or the name of an installed skill."),
     "inputSchema": {"type": "object", "properties": {
         "id": dict(_STR, description="result id, URL, or installed skill name"),
         "path": dict(_STR, description="path inside the repo, if the id is a bare repo")},
         "required": ["id"]}},
    {"name": "skills_install",
     "description": ("Install a skill into this host's catalog as a SKILL.md document. "
                     "Nothing is executed — a skill is instructions, never code. Use "
                     "all=true to take every skill in a repo (most packs hold several)."),
     "inputSchema": {"type": "object", "properties": {
         "id": dict(_STR, description="result id from skills_search"),
         "path": dict(_STR, description="path inside the repo"),
         "name": dict(_STR, description="rename it in the catalog"),
         "all": dict(_BOOL, description="install every skill in the repo")},
         "required": ["id"]}},
    {"name": "skills_installed",
     "description": "The installed catalog — every skill on this host, as cards.",
     "inputSchema": {"type": "object", "properties": {
         "q": dict(_STR, description="filter by text"),
         "tag": dict(_STR, description="filter by tag")}}},
    {"name": "skills_doc",
     "description": ("The markdown of one skill — what you actually put in front of a "
                     "model. Works on an installed name or on any result id, in which "
                     "case it is fetched live and not installed."),
     "inputSchema": {"type": "object", "properties": {
         "name": dict(_STR, description="installed skill name, or a result id")},
         "required": ["name"]}},
    {"name": "skills_load",
     "description": ("Several skills at once, bodies included — the call an agent makes "
                     "at the start of a run to learn what it is allowed to do. Name them, "
                     "or filter with q; unfiltered it returns the whole catalog."),
     "inputSchema": {"type": "object", "properties": {
         "names": dict(_LIST, description="skill names to load"),
         "q": dict(_STR, description="load everything matching this instead")}}},
    {"name": "skills_write",
     "description": ("Author a skill here instead of finding one: a name, a description "
                     "and the markdown instructions. Saved in the same catalog and in the "
                     "same format as a scraped one."),
     "inputSchema": {"type": "object", "properties": {
         "name": _STR, "body": dict(_STR, description="the instructions, as markdown"),
         "description": dict(_STR, description="one line — when should a model reach for this"),
         "tools": dict(_LIST, description="tools this skill expects the agent to have"),
         "tags": _LIST},
         "required": ["name", "body"]}},
    {"name": "skills_remove",
     "description": "Remove an installed skill from the catalog.",
     "inputSchema": {"type": "object", "properties": {"name": _STR},
                     "required": ["name"]}},
    {"name": "skills_token",
     "description": ("Set (or clear) the GitHub token that unlocks code search and lifts "
                     "the anonymous rate limit. Stored 0600 off-tree, never in the module."),
     "inputSchema": {"type": "object", "properties": {
         "token": dict(_STR, description="a GitHub PAT; empty string clears it")}}},
]

WRITE_TOOLS = {"skills_install", "skills_write", "skills_remove", "skills_token"}

_market: Market = None


def market() -> Market:
    global _market
    if _market is None:
        _market = Market()
    return _market


def call_tool(name: str, args: Dict[str, Any]) -> Any:
    m = market()
    a = args or {}
    if name == "skills_search":
        return m.search(a.get("q", ""), a.get("sources"), int(a.get("limit") or 30),
                        bool(a.get("fresh")))
    if name == "skills_sources":
        return m.sources()
    if name == "skills_get":
        return m.get(a.get("id"), a.get("path"))
    if name == "skills_install":
        return m.install(a.get("id"), a.get("path"), a.get("name"),
                         all=bool(a.get("all")), who=a.get("who", ""))
    if name == "skills_installed":
        return m.installed(a.get("q", ""), a.get("tag"))
    if name == "skills_doc":
        return m.doc(a.get("name"))
    if name == "skills_load":
        return m.publish(a.get("names"), a.get("q", ""))
    if name == "skills_write":
        return m.write(a.get("name"), a.get("body"), a.get("description", ""),
                       a.get("tools"), a.get("tags"), who=a.get("who", ""))
    if name == "skills_remove":
        return m.remove(a.get("name"))
    if name == "skills_token":
        return m.token(a.get("token"))
    raise KeyError(f"unknown tool: {name}")


def info() -> Dict[str, Any]:
    return {"module": "skills", "protocol": PROTOCOL, "server": SERVER,
            "tools": [{"name": t["name"], "description": t["description"]} for t in TOOLS],
            "sources": SOURCE_IDS}


def rpc(req: Dict[str, Any], authorized: bool = True) -> Dict[str, Any]:
    """One JSON-RPC message in, one out. `None` for a notification."""
    rid = req.get("id")
    method = req.get("method")
    params = req.get("params") or {}

    def ok(result):
        return {"jsonrpc": "2.0", "id": rid, "result": result}

    def err(code, message):
        return {"jsonrpc": "2.0", "id": rid, "error": {"code": code, "message": message}}

    if method == "initialize":
        return ok({"protocolVersion": PROTOCOL, "serverInfo": SERVER,
                   "capabilities": {"tools": {"listChanged": False}}})
    if method in ("notifications/initialized", "notifications/cancelled"):
        return None
    if method == "ping":
        return ok({})
    if method == "tools/list":
        return ok({"tools": TOOLS})
    if method == "tools/call":
        name = params.get("name")
        args = params.get("arguments") or {}
        if name in WRITE_TOOLS and not authorized:
            return err(-32001, "installing, writing or removing a skill is gated — "
                               "call from this box or send the bearer token")
        try:
            result = call_tool(name, args)
        except KeyError as e:
            return err(-32601, str(e))
        except Exception as e:
            return ok({"content": [{"type": "text", "text": f"error: {e}"}],
                       "isError": True})
        text = json.dumps(result, indent=2, default=str)
        if len(text) > 200_000:
            text = text[:200_000] + "\n… truncated"
        return ok({"content": [{"type": "text", "text": text}]})
    return err(-32601, f"unknown method: {method}")
